### PR TITLE
test: use pytest-forked lib to run test in forked process

### PR DESCRIPTION
**Pull request by girishc13**
**Goals:**

- Run tests in forked process to avoid flakiness and reusing process state for subsequent Flow's. Shutting down a Flow leaves the module space in an inconsistent state so forking multiple Flow in pytest is not so useful.

### CODE REVIEW

**Review by JoanFM — CHANGES_REQUESTED** [review]
Apply same change to `cd.yml`

File: tests/unit/test_yamlparser.py [hunk 1]
```diff
@@ -120,6 +120,7 @@ def test_encoder_name_env_replace():
 
 
 def test_encoder_name_dict_replace():
+    os.environ['BE_TEST_NAME'] = 'hello123'
```
Comment (JoanFM) [inline tests/unit/test_yamlparser.py:123]: can u please use a fixture to set the env variable and reset afterwards? this is done in many other places

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/test_yamlparser.py [hunk 2]
```diff
@@ -113,13 +113,20 @@ def test_expand_env():
...
-def test_encoder_name_env_replace():
+@pytest.fixture
+def set_test_encoder_env_var():
+    os.environ['BE_TEST_NAME'] = 'hello123'
+    yield
+    del os.environ['BE_TEST_NAME']
+
+
+def test_encoder_name_env_replace(set_test_encoder_env_var):
     os.environ['BE_TEST_NAME'] = 'hello123'
```
Comment (JoanFM) [inline tests/unit/test_yamlparser.py:124]: u can remove this here, but is okey

**Review by JoanFM — APPROVED** [review]

File: tests/unit/test_yamlparser.py (hunk 2 above)
Comment (girishc13) [reply tests/unit/test_yamlparser.py:124]: Ah, let me fix it so it's more consistent.

**Comment by girishc13** [comment]
@JoanFM I haven't added the arguments to the `cd.yml` yet. One thing to also notice with this change is that the CI logs don't capture the logging of each individual tests. The logs are not available because each test run in its own process. Locally we can remove the `--forked` argument and obtain the logs but on CI we might need to do some commits in case we want a detailed log.

**Comment by JoanFM** [comment]
> @JoanFM I haven't added the arguments to the `cd.yml` yet. One thing to also notice with this change is that the CI logs don't capture the logging of each individual tests. The logs are not available because each test run in its own process. Locally we can remove the `--forked` argument and obtain the logs but on CI we might need to do some commits in case we want a detailed log.

if it guarantees non-flakyness, I would still pay the price, however there must be some way to solve this

**Comment by girishc13** [comment]
Unfortunately the pytest-forked plugin isn't currently maintained. This [open issue](https://github.com/pytest-dev/pytest-forked/issues/21) is causing the `tests/unit/orchestrate/pods/` to fail. I'm not sure of an alternative to this test.

**Comment by girishc13** [comment]
The whole test code for `test_pod.py` looks unstable on close inspection and logging. The same pid is being reused across all tests. I added some logging to check the pid of each `SlowFakeRuntime`. The first runtime `Pod@41979` runs throughout the test so I'm not sure what the tests are doing.

```
pytest --suppress-no-test-exit-code -v -s --ignore-glob='tests/integration/hub_usage/dummyhub*' tests/unit/orchestrate/pods/test_pod.py | grep -e "creating"
tests/unit/orchestrate/pods/test_pod.py::test_pod_runtime_env_setting DEBUG  Pod@41979 --->creating new pod...                                                                                      [10/10/22 18:08:04]
tests/unit/orchestrate/pods/test_pod.py::test_gateway_args[grpc-GRPCGateway] DEBUG  gateway@41979 --->creating new pod...                                                                                  [10/10/22 18:08:04]
tests/unit/orchestrate/pods/test_pod.py::test_gateway_args[websocket-WebSocketGateway] DEBUG  gateway@41979 --->creating new pod...                                                                                  [10/10/22 18:08:04]
tests/unit/orchestrate/pods/test_pod.py::test_gateway_args[http-HTTPGateway] DEBUG  gateway@41979 --->creating new pod...                                                                                  [10/10/22 18:08:04]
tests/unit/orchestrate/pods/test_pod.py::test_gateway_runtimes[grpc-GRPCGateway] DEBUG  gateway@41979 --->creating new pod...                                                                                  [10/10/22 18:08:04]
tests/unit/orchestrate/pods/test_pod.py::test_gateway_runtimes[websocket-WebSocketGateway] DEBUG  gateway@41979 --->creating new pod...                                                                                  [10/10/22 18:08:04]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:52345 (Press CTRL+C to quit)
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
tests/unit/orchestrate/pods/test_pod.py::test_gateway_runtimes[http-HTTPGateway] DEBUG  gateway@41979 --->creating new pod...                                                                                  [10/10/22 18:08:05]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:50697 (Press CTRL+C to quit)
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
tests/unit/orchestrate/pods/test_pod.py::test_non_gateway_runtimes[WorkerRuntime] DEBUG  Pod@41979 --->creating new pod...                                                                                      [10/10/22 18:08:05]
tests/unit/orchestrate/pods/test_pod.py::test_non_gateway_runtimes[HeadRuntime] DEBUG  Pod@41979 --->creating new pod...                                                                                      [10/10/22 18:08:05]
tests/unit/orchestrate/pods/test_pod.py::test_failing_executor DEBUG  Pod@41979 --->creating new pod...                                                                                      [10/10/22 18:08:06]
tests/unit/orchestrate/pods/test_pod.py::test_failing_gateway_runtimes[grpc-GRPCGateway] DEBUG  gateway@41979 --->creating new pod...                                                                                  [10/10/22 18:08:06]
tests/unit/orchestrate/pods/test_pod.py::test_failing_gateway_runtimes[websocket-WebSocketGateway] DEBUG  gateway@41979 --->creating new pod...                                                                                  [10/10/22 18:08:06]
tests/unit/orchestrate/pods/test_pod.py::test_failing_gateway_runtimes[http-HTTPGateway] DEBUG  gateway@41979 --->creating new pod...                                                                                  [10/10/22 18:08:06]
tests/unit/orchestrate/pods/test_pod.py::test_failing_head DEBUG  Pod@41979 --->creating new pod...                                                                                      [10/10/22 18:08:06]
tests/unit/orchestrate/pods/test_pod.py::test_close_before_start DEBUG  Pod@41979 --->creating new pod...                                                                                      [10/10/22 18:08:06]
tests/unit/orchestrate/pods/test_pod.py::test_close_before_start_slow_enter DEBUG  Pod@41979 --->creating new pod...                                                                                      [10/10/22 18:08:06]
```

**Comment by girishc13** [comment]
Some tests might still be flaky due to the incompatibility with the `purest-timeout` feature.

**Comment by JoanFM** [comment]
> Unfortunately the pytest-forked plugin isn't currently maintained. This [open issue](https://github.com/pytest-dev/pytest-forked/issues/21) is causing the `tests/unit/orchestrate/pods/` to fail. I'm not sure of an alternative to this test.

if it is not mantained, I would remove this

**Comment by girishc13** [comment]
The pytest-forked library currently doesn't have a maintainer and the open issue with `pytest-timeout` prevents us form investing further time.

**Pull request closed without merging**